common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -1018,51 +1019,57 @@ private static boolean isRoot0() {
         return false;
     }
 
-    @SuppressWarnings("LoopStatementThatDoesntLoop")
     private static int javaVersion0() {
-        int javaVersion;
+        final int majorVersion;
 
-        // Not really a loop
-        for (;;) {
-            // Android
-            if (isAndroid()) {
-                javaVersion = 6;
-                break;
-            }
+        if (isAndroid()) {
+            majorVersion = 6;
+        } else {
+            majorVersion = majorVersionFromJavaSpecificationVersion();
+        }
 
-            try {
-                Method getVersion = java.lang.Runtime.class.getMethod("version");
-                Object version = getVersion.invoke(null);
-                javaVersion = (Integer) version.getClass().getMethod("major").invoke(version);
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        logger.debug("Java version: {}", majorVersion);
 
-            try {
-                Class.forName("java.time.Clock", false, getClassLoader(Object.class));
-                javaVersion = 8;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        return majorVersion;
+    }
 
-            try {
-                Class.forName("java.util.concurrent.LinkedTransferQueue", false, getClassLoader(BlockingQueue.class));
-                javaVersion = 7;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+    static int majorVersionFromJavaSpecificationVersion() {
+        try {
+            final String javaSpecVersion = AccessController.doPrivileged(new PrivilegedAction<String>() {
+                @Override
+                public String run() {
+                    return System.getProperty("java.specification.version");
+                }
+            });
+            return majorVersion(javaSpecVersion);
+        } catch (SecurityException e) {
+            logger.debug("security exception while reading java.specification.version", e);
+            return 6;
+        }
+    }
 
-            javaVersion = 6;
-            break;
+    static int majorVersion(final String javaSpecVersion) {
+        final String[] components = javaSpecVersion.split("\\.");
+        final int[] version = new int[components.length];
+        for (int i = 0; i < components.length; i++) {
+            version[i] = Integer.parseInt(components[i]);
         }
 
-        if (logger.isDebugEnabled()) {
-            logger.debug("Java version: {}", javaVersion);
+        // Java 8 can return both 1.8 and 8, we rewrite the former to the latter
```

Comment:
@johnou Thanks for looking. After Project Verona, JDK 9 can only return "9*", it's only early release builds of JDK 9 before Project Verona was integrated that can return "1.9*". No one should be using those builds any longer :smile:. Yet, the method used here works on both "1.9*" and "9*" (see the unit tests added in this PR). However, your comment made me realize that this canonicalization is not needed here at all. I introduced [similar code in Elasticsearch](https://github.com/elastic/elasticsearch/blob/f79fb4ada7995c11874346787db276b6edafaa8f/core/src/main/java/org/elasticsearch/bootstrap/JavaVersion.java#L35-L44) where the canonicalization is needed because it's used there to compare versions. For parsing, we don't need this. I will push a commit to remove.